<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Single-cell differential expression analysis with FLASHMM • FLASHMM</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Single-cell differential expression analysis with FLASHMM">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLASHMM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/FLASHMM-vignette.html">Single-cell differential expression analysis with FLASHMM</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/BaderLab/FLASHMM/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Single-cell differential expression analysis with FLASHMM</h1>
                        <h4 data-toc-skip class="author">Changjiang Xu,
Delaram Pouyabahar, Veronique Voisin, and Gary D. Bader</h4>
            
            <h4 data-toc-skip class="date">January 07, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BaderLab/FLASHMM/blob/HEAD/vignettes/FLASHMM-vignette.Rmd"><code>vignettes/FLASHMM-vignette.Rmd</code></a></small>
      <div class="d-none name"><code>FLASHMM-vignette.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      FLASHMM is a package for analysis of single-cell differential
      expression (DE) using a linear mixed-effects model (LMM). The
      mixed-effects models have become a powerful tool in single-cell
      studies due to their ability to model intra-subject correlation
      and inter-subject variability. The classic LMM estimation method
      faces limitations of speed and computer memory in analysis of
      large scale scRNA-seq data. The package FLASHMM provides a fast
      and scalable method to address the scalability of the LMM
      estimation. FLASHMM is fast and requires less computer memory to
      fit the LMM. This vignette describes the method for LMM estimation
      and inference and the R functions for implementing the method, and
      demonstrates the use of the package via an example.
    </div>
    
<div class="section level2">
<h2 id="method">Method<a class="anchor" aria-label="anchor" href="#method"></a>
</h2>
<div class="section level3">
<h3 id="lmm-estimation-and-inference">LMM estimation and inference<a class="anchor" aria-label="anchor" href="#lmm-estimation-and-inference"></a>
</h3>
<p>Consider the linear mixed-effects model (LMM) as expressed below
<span class="citation">(<a href="#ref-Searle2006">Searle, Casella, and
McCulloch 2006</a>)</span>
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>#</mi><mi>e</mi><mi>q</mi><mo>:</mo><mi>l</mi><mi>m</mi><mi>m</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>y</mi><mo>=</mo><mi>X</mi><mi>β</mi><mo>+</mo><mi>Z</mi><mi>b</mi><mo>+</mo><mi>ϵ</mi><mo>,</mo></mrow><annotation encoding="application/x-tex">\begin{equation}
(\#eq:lmm)
y = X\beta + Zb + \epsilon,
\end{equation}</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>y</mi><annotation encoding="application/x-tex">y</annotation></semantics></math>
is an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">n\times 1</annotation></semantics></math>
vector of observed responses,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
is an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">n\times p</annotation></semantics></math>
design matrix for fixed effects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>β</mi><annotation encoding="application/x-tex">\beta</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Z</mi><annotation encoding="application/x-tex">Z</annotation></semantics></math>
is an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><mi>q</mi></mrow><annotation encoding="application/x-tex">n\times q</annotation></semantics></math>
design matrix for random effects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>b</mi><annotation encoding="application/x-tex">b</annotation></semantics></math>,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>ϵ</mi><annotation encoding="application/x-tex">\epsilon</annotation></semantics></math>
is an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">n\times 1</annotation></semantics></math>
vector of residual errors. The term of random effects may be a
combination of various components,
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Z</mi><mi>b</mi><mo>=</mo><msub><mi>Z</mi><mn>1</mn></msub><msub><mi>b</mi><mn>1</mn></msub><mo>+</mo><mi>⋯</mi><mo>+</mo><msub><mi>Z</mi><mi>K</mi></msub><msub><mi>b</mi><mi>K</mi></msub><mo>,</mo></mrow><annotation encoding="application/x-tex">
Zb = Z_1 b_1 + \cdots + Z_K b_K,
</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Z</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>Z</mi><mn>1</mn></msub><mo>,</mo><mi>…</mi><mo>,</mo><msub><mi>Z</mi><mi>K</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">Z=[Z_1,\ldots,Z_K]</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>b</mi><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>b</mi><mn>1</mn><mi>T</mi></msubsup><mo>,</mo><mi>…</mi><mo>,</mo><msubsup><mi>b</mi><mi>K</mi><mi>T</mi></msubsup><mo stretchy="true" form="postfix">]</mo></mrow><mi>T</mi></msup></mrow><annotation encoding="application/x-tex">b=[b^T_1,\ldots,b^T_K]^T</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
is the number of random-effect components, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Z</mi><mi>k</mi></msub><annotation encoding="application/x-tex">Z_k</annotation></semantics></math>
is an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><msub><mi>q</mi><mi>k</mi></msub></mrow><annotation encoding="application/x-tex">n\times q_k</annotation></semantics></math>
design matrix for the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>-th
random-effect component. The superscript
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>T</mi><annotation encoding="application/x-tex">T</annotation></semantics></math>
denotes a transpose of vector or matrix. The basic assumptions are as
follows:</p>
<ol style="list-style-type: decimal">
<li>The design matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
is of full rank, satisfying conditions of estimability for the
parameters;</li>
<li>The random vectors
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>b</mi><mi>k</mi></msub><annotation encoding="application/x-tex">b_k</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>ϵ</mi><annotation encoding="application/x-tex">\epsilon</annotation></semantics></math>
are independent and follow a normal distribution,
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>b</mi><mi>k</mi></msub><mo>∼</mo><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>𝟎</mn><mo>,</mo><msubsup><mi>σ</mi><mi>k</mi><mn>2</mn></msubsup><msub><mi>I</mi><msub><mi>q</mi><mi>k</mi></msub></msub><mo stretchy="true" form="postfix">)</mo></mrow><mspace width="1.0em"></mspace><mtext mathvariant="normal">and</mtext><mspace width="1.0em"></mspace><mi>ϵ</mi><mo>∼</mo><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>𝟎</mn><mo>,</mo><msup><mi>σ</mi><mn>2</mn></msup><msub><mi>I</mi><mi>n</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mi>.</mi></mrow><annotation encoding="application/x-tex">b_k \sim N(\mathbf{0}, \sigma^2_k I_{q_k})\quad\text{and}\quad\epsilon \sim N(\mathbf{0}, \sigma^2I_n).</annotation></semantics></math>
</li>
</ol>
<p>Here
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>𝟎</mn><annotation encoding="application/x-tex">\mathbf{0}</annotation></semantics></math>
is a vector or matrix of zero elements,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mi>n</mi></msub><annotation encoding="application/x-tex">I_n</annotation></semantics></math>
is an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><mi>n</mi></mrow><annotation encoding="application/x-tex">n\times n</annotation></semantics></math>
identity matrix, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><mi>k</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_k</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>σ</mi><mn>2</mn></msup><annotation encoding="application/x-tex">\sigma^2</annotation></semantics></math>
are unknown parameters, called variance components.</p>
<p><span class="citation">Hartley and Rao (<a href="#ref-HartleyRao1967">1967</a>)</span> developed the maximum
likelihood estimation (MLE) method for estimating the LMM parameters
(fixed effects and variance components). <span class="citation">Patterson and Thompson (<a href="#ref-PattersonThompson1971">1971</a>)</span> proposed a modified
maximum likelihood procedure which partitions the data into two mutually
uncorrelated parts, one being free of the fixed effects used for
estimating variance components, called restricted maximum likelihood
(REML) estimator. The REML estimator is unbiased. The MLE of variance
components is biased in general. Both methods are asymptotically
identical for estimating variance components. With variance components,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>θ</mi><annotation encoding="application/x-tex">\theta</annotation></semantics></math>,
estimated, the fixed effects estimated by either MLE or REML are given
as follows
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover><mi>β</mi><mo accent="true">̂</mo></mover><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>X</mi><mi>T</mi></msup><msubsup><mi>V</mi><mi>θ</mi><mrow><mi>−</mi><mn>1</mn></mrow></msubsup><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mi>−</mi><mn>1</mn></mrow></msup><msup><mi>X</mi><mi>T</mi></msup><msubsup><mi>V</mi><mi>θ</mi><mrow><mi>−</mi><mn>1</mn></mrow></msubsup><mi>y</mi><mo>,</mo></mrow><annotation encoding="application/x-tex">
\hat\beta = (X^TV_{\theta}^{-1}X )^{-1}X^TV_{\theta}^{-1}y,
</annotation></semantics></math> with covariance matrix
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>v</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>β</mi><mo accent="true">̂</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>X</mi><mi>T</mi></msup><msubsup><mi>V</mi><mi>θ</mi><mrow><mi>−</mi><mn>1</mn></mrow></msubsup><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mi>−</mi><mn>1</mn></mrow></msup><mo>,</mo></mrow><annotation encoding="application/x-tex">var(\hat\beta) = (X^TV_{\theta}^{-1}X)^{-1},</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>θ</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>θ</mi><mn>0</mn></msub><mo>,</mo><msub><mi>θ</mi><mn>1</mn></msub><mo>,</mo><mi>…</mi><mo>,</mo><msub><mi>θ</mi><mi>K</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\theta = (\theta_0, \theta_1,\ldots, \theta_K)</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>θ</mi><mn>0</mn></msub><mo>=</mo><msup><mi>σ</mi><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">\theta_0 = \sigma^2</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>θ</mi><mi>k</mi></msub><mo>=</mo><msubsup><mi>σ</mi><mi>k</mi><mn>2</mn></msubsup></mrow><annotation encoding="application/x-tex">\theta_k = \sigma^2_k</annotation></semantics></math>,
and
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>V</mi><mi>θ</mi></msub><mo>=</mo><msub><mi>θ</mi><mn>0</mn></msub><msub><mi>I</mi><mi>n</mi></msub><mo>+</mo><msub><mi>θ</mi><mn>1</mn></msub><msub><mi>Z</mi><mn>1</mn></msub><msubsup><mi>Z</mi><mn>1</mn><mi>T</mi></msubsup><mo>+</mo><mi>…</mi><mo>+</mo><msub><mi>θ</mi><mi>K</mi></msub><msub><mi>Z</mi><mi>K</mi></msub><msubsup><mi>Z</mi><mi>K</mi><mi>T</mi></msubsup><mi>.</mi></mrow><annotation encoding="application/x-tex">V_{\theta} = \theta_0 I_n + \theta_1 Z_1Z_1^T + \ldots + \theta_K Z_KZ_K^T.</annotation></semantics></math></p>
<p>Estimating the variance components by either MLE or REML is a
difficult numerical problem. Various iterative methods based on the log
likelihood, called gradient methods, have been proposed <span class="citation">(<a href="#ref-Searle2006">Searle, Casella, and
McCulloch 2006</a>)</span>. The gradient methods are represented by the
iteration equation
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>#</mi><mi>e</mi><mi>q</mi><mo>:</mo><mi>g</mi><mi>r</mi><mi>a</mi><mi>d</mi><mi>i</mi><mi>e</mi><mi>n</mi><mi>t</mi><mo stretchy="true" form="postfix">)</mo></mrow><msup><mi>θ</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>+</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></msup><mo>=</mo><msup><mi>θ</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></msup><mo>+</mo><mi>Γ</mi><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>θ</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></msup><mo stretchy="true" form="postfix">)</mo></mrow><mfrac><mrow><mi>∂</mi><mi>l</mi><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>θ</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></msup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><mi>∂</mi><mi>θ</mi></mrow></mfrac><mo>,</mo></mrow><annotation encoding="application/x-tex">\begin{equation}
(\#eq:gradient)
\theta^{(i+1)} = \theta^{(i)} + \Gamma(\theta^{(i)})\frac{\partial l(\theta^{(i)})}{\partial\theta},
\end{equation}</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>∂</mi><mi>l</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>θ</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>/</mi><mi>∂</mi><mi>θ</mi></mrow><annotation encoding="application/x-tex">\partial l(\theta)/\partial\theta</annotation></semantics></math>
is the gradient of the log likelihood function, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Γ</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>θ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\Gamma(\theta)</annotation></semantics></math>
is a modifier matrix of the gradient direction, which can be specified
by Newton–Raphson, Fisher scoring or average information, see
Supplementary Materials of the manuscript for the details.</p>
<p>The hypotheses for testing fixed effects and variance components can
be respectively defined as
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mrow><mn>0</mn><mo>,</mo><mi>i</mi></mrow></msub><mo>:</mo><msub><mi>β</mi><mi>i</mi></msub><mo>=</mo><mn>0</mn><mspace width="0.222em"></mspace><mspace width="0.222em"></mspace><mtext mathvariant="normal">versus</mtext><mspace width="0.222em"></mspace><mspace width="0.222em"></mspace><msub><mi>H</mi><mrow><mn>1</mn><mo>,</mo><mi>i</mi></mrow></msub><mo>:</mo><msub><mi>β</mi><mi>i</mi></msub><mo>≠</mo><mn>0</mn><mo>,</mo></mrow><annotation encoding="application/x-tex">
H_{0, i}: \beta_i = 0 ~~\text{versus}~~H_{1,i}: \beta_i\ne 0,
</annotation></semantics></math><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mrow><mn>0</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>:</mo><msub><mi>θ</mi><mi>k</mi></msub><mo>≤</mo><mn>0</mn><mspace width="0.222em"></mspace><mspace width="0.222em"></mspace><mtext mathvariant="normal">versus</mtext><mspace width="0.222em"></mspace><mspace width="0.222em"></mspace><msub><mi>H</mi><mrow><mn>1</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>:</mo><msub><mi>θ</mi><mi>k</mi></msub><mo>&gt;</mo><mn>0</mn><mo>,</mo></mrow><annotation encoding="application/x-tex">
H_{0, k}: \theta_k \leq 0 ~~\text{versus}~~H_{1,k}: \theta_k &gt; 0,
</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>θ</mi><mi>k</mi></msub><annotation encoding="application/x-tex">\theta_k</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mo>=</mo><mn>1</mn><mo>,</mo><mi>…</mi><mo>,</mo><mi>K</mi></mrow><annotation encoding="application/x-tex">k=1, \ldots, K</annotation></semantics></math>,
represent the parameters of variance components
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><mi>k</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_k</annotation></semantics></math>
but are allowed to be negative. The lower boundary of the parameters,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>θ</mi><annotation encoding="application/x-tex">\theta</annotation></semantics></math>,
can be the negative value such that the variance-covariance matrix,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>θ</mi></msub><annotation encoding="application/x-tex">V_{\theta}</annotation></semantics></math>,
remains definable (positive definite). Note that the negative lower
boundary exists and can be less than
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>−</mi><msup><mi>σ</mi><mn>2</mn></msup><mi>/</mi><msub><mi>λ</mi><mrow><mi>m</mi><mi>a</mi><mi>x</mi></mrow></msub></mrow><annotation encoding="application/x-tex">- \sigma^2/\lambda_{max}</annotation></semantics></math>,
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>λ</mi><mrow><mi>m</mi><mi>a</mi><mi>x</mi></mrow></msub><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\lambda_{max} &gt; 0</annotation></semantics></math>,
is the largest singular value of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Z</mi><msup><mi>Z</mi><mi>T</mi></msup></mrow><annotation encoding="application/x-tex">ZZ^T</annotation></semantics></math>.
If
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>θ</mi><mi>k</mi></msub><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\theta_k &gt; 0</annotation></semantics></math>,
then
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>σ</mi><mi>k</mi><mn>2</mn></msubsup><mo>=</mo><msub><mi>θ</mi><mi>k</mi></msub></mrow><annotation encoding="application/x-tex">\sigma_k^2 = \theta_k</annotation></semantics></math>
is definable and the mixed model is well-specified. Otherwise, if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>θ</mi><mi>k</mi></msub><mo>≤</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\theta_k \le 0</annotation></semantics></math>,
the mixed model is miss-specified and no random effects are needed.</p>
<p>Allowing the parameters of variance components to take negative value
avoids the zero boundary problem in hypothesis testing for the variance
components. Consequently, the asymptotic properties of the maximum
likelihood estimation hold under regularity conditions, which enables us
to use z-statistics or t-statistics for hypothesis testing of fixed
effects and variance components. The t-statistics for fixed effects are
given by
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>#</mi><mi>e</mi><mi>q</mi><mo>:</mo><mi>t</mi><mi>c</mi><mi>o</mi><mi>e</mi><mi>f</mi><mo stretchy="true" form="postfix">)</mo></mrow><msub><mi>T</mi><mi>i</mi></msub><mo>=</mo><mfrac><msub><mover><mi>β</mi><mo accent="true">̂</mo></mover><mi>i</mi></msub><msqrt><mrow><mi>v</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mover><mi>β</mi><mo accent="true">̂</mo></mover><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msqrt></mfrac><mo>=</mo><mfrac><msub><mover><mi>β</mi><mo accent="true">̂</mo></mover><mi>i</mi></msub><msqrt><mrow><mi>v</mi><mi>a</mi><mi>r</mi><msub><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>β</mi><mo accent="true">̂</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mi>i</mi><mi>i</mi></mrow></msub></mrow></msqrt></mfrac><mspace width="0.222em"></mspace><mo>∼</mo><mspace width="0.222em"></mspace><mi>t</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>n</mi><mo>−</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>.</mi></mrow><annotation encoding="application/x-tex">\begin{equation}
(\#eq:tcoef)
T_i = \frac{\hat\beta_i}{\sqrt{var(\hat\beta_i)}} = \frac{\hat\beta_i}{\sqrt{var(\hat\beta)_{ii}}} ~\sim ~t(n - p).
\end{equation}</annotation></semantics></math> The t-statistic for a
contrast, a linear combination of the estimated fixed effects,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>c</mi><mi>T</mi></msup><mover><mi>β</mi><mo accent="true">̂</mo></mover></mrow><annotation encoding="application/x-tex">c^T\hat\beta</annotation></semantics></math>,
is
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>#</mi><mi>e</mi><mi>q</mi><mo>:</mo><mi>t</mi><mi>c</mi><mi>o</mi><mi>n</mi><mi>t</mi><mi>r</mi><mi>a</mi><mi>s</mi><mi>t</mi><mo stretchy="true" form="postfix">)</mo></mrow><msub><mi>T</mi><mi>c</mi></msub><mo>=</mo><mfrac><mrow><msup><mi>c</mi><mi>T</mi></msup><mover><mi>β</mi><mo accent="true">̂</mo></mover></mrow><msqrt><mrow><msup><mi>c</mi><mi>T</mi></msup><mi>v</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>β</mi><mo accent="true">̂</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mi>c</mi></mrow></msqrt></mfrac><mo>∼</mo><mi>t</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>n</mi><mo>−</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>.</mi></mrow><annotation encoding="application/x-tex">\begin{equation}
(\#eq:tcontrast)
T_c = \frac{c^T\hat\beta}{\sqrt{c^Tvar(\hat\beta) c}} \sim t(n-p).
\end{equation}</annotation></semantics></math> The z-statistics for the
parameters of variance components are given by
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>#</mi><mi>e</mi><mi>q</mi><mo>:</mo><mi>z</mi><mi>v</mi><mi>a</mi><mi>r</mi><mi>c</mi><mi>o</mi><mi>m</mi><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow><msub><mi>Z</mi><mi>k</mi></msub><mo>=</mo><mfrac><msub><mover><mi>θ</mi><mo accent="true">̂</mo></mover><mi>k</mi></msub><msqrt><msub><mrow><mo stretchy="true" form="prefix">[</mo><mi>I</mi><msup><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>θ</mi><mo accent="true">̂</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mi>−</mi><mn>1</mn></mrow></msup><mo stretchy="true" form="postfix">]</mo></mrow><mrow><mi>k</mi><mi>k</mi></mrow></msub></msqrt></mfrac><mo>∼</mo><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo></mrow><annotation encoding="application/x-tex">\begin{equation}
(\#eq:zvarcomp)
Z_k = \frac{\hat\theta_k}{\sqrt{[I(\hat\theta)^{-1}]_{kk}}} \sim N(0, 1),
\end{equation}</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>I</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>θ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">I(\theta)</annotation></semantics></math>
is the Fisher information matrix.</p>
</div>
<div class="section level3">
<h3 id="fast-and-scalable-algorithm">Fast and scalable algorithm<a class="anchor" aria-label="anchor" href="#fast-and-scalable-algorithm"></a>
</h3>
<p>We developed a summary statistics based algorithm for implementing
the gradient method @ref(eq:gradient). Let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mi>X</mi></mrow><annotation encoding="application/x-tex">XX</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mi>Y</mi></mrow><annotation encoding="application/x-tex">XY</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Z</mi><mi>X</mi></mrow><annotation encoding="application/x-tex">ZX</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Z</mi><mi>Y</mi></mrow><annotation encoding="application/x-tex">ZY</annotation></semantics></math>,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Z</mi><mi>Z</mi></mrow><annotation encoding="application/x-tex">ZZ</annotation></semantics></math>
denote a matrix, respectively, which define the summary statistics that
are computed from cell-level data
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Z</mi><annotation encoding="application/x-tex">Z</annotation></semantics></math>
as follows:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>#</mi><mi>e</mi><mi>q</mi><mo>:</mo><mi>s</mi><mi>d</mi><mi>a</mi><mi>t</mi><mi>a</mi><mo stretchy="true" form="postfix">)</mo></mrow><mtable><mtr><mtd columnalign="left" style="text-align: left"><mi>X</mi><mi>X</mi><mo>=</mo><msup><mi>X</mi><mi>T</mi></msup><mi>X</mi><mo>,</mo><mspace width="0.222em"></mspace><mi>X</mi><mi>Y</mi><mo>=</mo><msup><mi>X</mi><mi>T</mi></msup><msup><mi>Y</mi><mi>T</mi></msup><mo>,</mo></mtd></mtr><mtr><mtd columnalign="left" style="text-align: left"><mi>Z</mi><mi>X</mi><mo>=</mo><msup><mi>Z</mi><mi>T</mi></msup><mi>X</mi><mo>,</mo><mspace width="0.222em"></mspace><mi>Z</mi><mi>Y</mi><mo>=</mo><msup><mi>Z</mi><mi>T</mi></msup><msup><mi>Y</mi><mi>T</mi></msup><mo>,</mo><mspace width="0.222em"></mspace><mi>Z</mi><mi>Z</mi><mo>=</mo><msup><mi>Z</mi><mi>T</mi></msup><mi>Z</mi><mo>,</mo></mtd></mtr><mtr><mtd columnalign="left" style="text-align: left"><mi>Y</mi><mi>n</mi><mi>o</mi><mi>r</mi><mi>m</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>y</mi><mn>1</mn></msub><msubsup><mi>y</mi><mn>1</mn><mi>T</mi></msubsup><mo>,</mo><mi>…</mi><mo>,</mo><msub><mi>y</mi><mi>m</mi></msub><msubsup><mi>y</mi><mi>m</mi><mi>T</mi></msubsup><mo stretchy="true" form="postfix">]</mo></mrow><mo>,</mo></mtd></mtr></mtable></mrow><annotation encoding="application/x-tex">\begin{equation}
(\#eq:sdata)
\begin{array}{l}
XX = X^TX, ~XY = X^TY^T,\\
ZX = Z^TX, ~ZY = Z^TY^T, ~ZZ = Z^TZ,\\
Ynorm = [y_1y_1^T, \ldots, y_my_m^T],
\end{array}
\end{equation}</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Y</mi><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>y</mi><mn>1</mn><mi>T</mi></msubsup><mo>,</mo><mi>…</mi><mo>,</mo><msubsup><mi>y</mi><mi>m</mi><mi>T</mi></msubsup><mo stretchy="true" form="postfix">]</mo></mrow><mi>T</mi></msup></mrow><annotation encoding="application/x-tex">Y = [y_1^T, \ldots, y_m^T]^T</annotation></semantics></math>
is a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>-by-<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
matrix of gene expression profile with each row
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>y</mi><mi>i</mi></msub><annotation encoding="application/x-tex">y_i</annotation></semantics></math>
corresponding to the expression of gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mo>=</mo><mn>1</mn><mo>,</mo><mi>…</mi><mo>,</mo><mi>m</mi></mrow><annotation encoding="application/x-tex">i=1,\ldots,m</annotation></semantics></math>.
Once the summary statistics are precomputed, the summary-level data
based algorithm has a complexity of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>m</mi><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>p</mi><mn>3</mn></msup><mo>+</mo><msup><mi>q</mi><mn>3</mn></msup><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(m(p^3 + q^3))</annotation></semantics></math>,
which doesn’t depend on number of cells (sample size
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>).
In single-cell DE analysis, the numbers of fixed and random effects,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>q</mi><annotation encoding="application/x-tex">q</annotation></semantics></math>,
are relatively small. Therefore, the algorithm is fast and scalable, and
requires less computer memory. See Supplemental Materials of the
manuscript for details.</p>
</div>
</div>
<div class="section level2">
<h2 id="flashmm">FLASHMM<a class="anchor" aria-label="anchor" href="#flashmm"></a>
</h2>
<p>FLASHMM package provides two functions, lmm and lmmfit, for fitting
LMM. The lmm function uses summary statistics as arguments. The lmmfit
function is a wrapper function of lmm, which directly uses cell-level
data and computes the summary statistics inside the function. The lmmfit
function is simple to be operated but it has a limitation of memory use.
For large scale data, we should use lmm instead of lmmfit. That is, we
precompute the summary-level data by @ref(eq:sdata) or sslmm function in
FLASHMM package, and then use lmm function to fit LMM. FLASHMM provides
the lmmtest function to perform statistical test on the fixed effects
and the contrasts of fixed effects.</p>
<p>In summary, FLASHMM package provides the following functions.</p>
<ul>
<li>lmm: fit LMM using summary-level data.</li>
<li>lmmfit: fit LMM using cell-level data.</li>
<li>lmmtest: perform statistical tests on the fixed effects and their
contrasts.</li>
<li>sslmm: compute the summary-level data using cell-level data.</li>
<li>simuRNAseq: simulate multi-sample multi-cell-type scRNA-seq dataset
based on a negative binomial distribution.</li>
</ul>
<div class="section level3">
<h3 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Install FLASHMM from Github.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"https://github.com/Baderlab/FLASHMM"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Load the package.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BaderLab/FLASHMM">FLASHMM</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="simulating-a-scrna-seq-dataset-by-simurnaseq">Simulating a scRNA-seq dataset by simuRNAseq<a class="anchor" aria-label="anchor" href="#simulating-a-scrna-seq-dataset-by-simurnaseq"></a>
</h4>
<p>Simulate a multi-sample multi-cell-cluster scRNA-seq dataset
comprising 25 samples and 4 clusters (cell-types) with 2 treatments.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2412</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simuRNAseq.html">simuRNAseq</a></span><span class="op">(</span>nGenes <span class="op">=</span> <span class="fl">50</span>, nCells <span class="op">=</span> <span class="fl">1000</span>, </span>
<span>                  nsam <span class="op">=</span> <span class="fl">25</span>, ncls <span class="op">=</span> <span class="fl">4</span>, ntrt <span class="op">=</span> <span class="fl">2</span>, nDEgenes <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "ref.mean.dispersion" "metadata"            "counts"             </span></span>
<span><span class="co">#&gt; [4] "DEgenes"             "treatment"</span></span>
<span><span class="co">##</span></span>
<span></span>
<span><span class="co">#counts and meta data</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">dat</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">dat</span><span class="op">$</span><span class="va">metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="de-analysis-using-lmm">DE analysis using LMM<a class="anchor" aria-label="anchor" href="#de-analysis-using-lmm"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##(1) Model design</span></span>
<span><span class="co">##Y: gene expression profile (log-transformed counts)</span></span>
<span><span class="co">##X: design matrix for fixed effects</span></span>
<span><span class="co">##Z: design matrix for random effects</span></span>
<span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> </span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">libsize</span><span class="op">)</span> <span class="op">+</span> <span class="va">cls</span> <span class="op">+</span> <span class="va">cls</span><span class="op">:</span><span class="va">trt</span>, data <span class="op">=</span> <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="va">Z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">sam</span>, data <span class="op">=</span> <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span></span>
<span><span class="co">##</span></span>
<span></span>
<span><span class="co">##(2) LMM fitting</span></span>
<span><span class="co">##Fit LMM by lmmfit using cell-level data.</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lmmfit.html">lmmfit</a></span><span class="op">(</span><span class="va">Y</span>, <span class="va">X</span>, <span class="va">Z</span>, d <span class="op">=</span> <span class="va">d</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##Fit LMM by lmm using summary-level data computed as follows.</span></span>
<span><span class="co">##- Computing summary statistics</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">XX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">X</span>; <span class="va">XY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">Y</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">ZX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">X</span>; <span class="va">ZY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">Y</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">Z</span><span class="op">)</span>; <span class="va">ZZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">Z</span></span>
<span><span class="va">Ynorm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">Y</span><span class="op">*</span><span class="va">Y</span><span class="op">)</span></span>
<span><span class="co">##- Fitting LMM</span></span>
<span><span class="va">fitss</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lmm.html">lmm</a></span><span class="op">(</span><span class="va">XX</span>, <span class="va">XY</span>, <span class="va">ZX</span>, <span class="va">ZY</span>, <span class="va">ZZ</span>, Ynorm <span class="op">=</span> <span class="va">Ynorm</span>, n <span class="op">=</span> <span class="va">n</span>, d <span class="op">=</span> <span class="va">d</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">fitss</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">##</span></span>
<span></span>
<span><span class="co">##Fit LMM by lmm using summary-level data computed by sslmm.</span></span>
<span><span class="co">##- Computing summary statistics</span></span>
<span><span class="va">ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sslmm.html">sslmm</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span>, <span class="va">Z</span><span class="op">)</span></span>
<span><span class="co">##- Fitting LMM</span></span>
<span><span class="va">fitss</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lmm.html">lmm</a></span><span class="op">(</span>summary.stats <span class="op">=</span> <span class="va">ss</span>, d <span class="op">=</span> <span class="va">d</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">fitss</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">##</span></span>
<span></span>
<span><span class="co">##(3) Hypothesis tests</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lmmtest.html">lmmtest</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span><span class="co">#head(test)</span></span>
<span></span>
<span><span class="co">##t-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">t</span><span class="op">)</span> <span class="op">==</span> <span class="va">test</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"_t"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="va">fit</span><span class="op">$</span><span class="va">t</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                   Gene1      Gene2      Gene3      Gene4      Gene5</span></span>
<span><span class="co">#&gt; log(libsize)  3.5564382  6.1353128  3.6098989  4.1239445  2.8179995</span></span>
<span><span class="co">#&gt; clsC1        -2.4158938 -4.8697164 -2.3122737 -2.6709945 -1.3220206</span></span>
<span><span class="co">#&gt; clsC2        -2.5957105 -4.9918921 -2.2467924 -2.5108410 -1.2387859</span></span>
<span><span class="co">#&gt; clsC3        -2.5576263 -5.0249733 -2.1484996 -2.5856150 -1.2535998</span></span>
<span><span class="co">#&gt; clsC4        -2.4466344 -5.0065480 -2.2202673 -2.6983076 -1.2058758</span></span>
<span><span class="co">#&gt; clsC1:trtB    0.8697778  0.3659413  0.1515357  0.9707563 -0.9577914</span></span>
<span><span class="co">#&gt; clsC2:trtB    2.0700456  1.0976568 -0.1112106  0.7423556 -0.5997369</span></span>
<span><span class="co">#&gt; clsC3:trtB    1.5066385  1.5001771 -0.3659261  0.8883383 -1.0410003</span></span>
<span><span class="co">#&gt; clsC4:trtB   -0.3263183  1.6810047 -0.4559326  0.6462646 -1.7515738</span></span>
<span><span class="co">##</span></span>
<span></span>
<span><span class="co">##p-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">p</span><span class="op">)</span> <span class="op">==</span> <span class="va">test</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"_p"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="va">fit</span><span class="op">$</span><span class="va">p</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                     Gene1        Gene2        Gene3        Gene4       Gene5</span></span>
<span><span class="co">#&gt; log(libsize) 0.0003936946 1.226867e-09 0.0003216502 4.036515e-05 0.004928509</span></span>
<span><span class="co">#&gt; clsC1        0.0158766072 1.300111e-06 0.0209667982 7.686618e-03 0.186466367</span></span>
<span><span class="co">#&gt; clsC2        0.0095791682 7.060831e-07 0.0248727531 1.220264e-02 0.215718133</span></span>
<span><span class="co">#&gt; clsC3        0.0106867912 5.971329e-07 0.0319158551 9.862323e-03 0.210283172</span></span>
<span><span class="co">#&gt; clsC4        0.0145925607 6.556356e-07 0.0266262016 7.087769e-03 0.228153191</span></span>
<span><span class="co">#&gt; clsC1:trtB   0.3846324624 7.144869e-01 0.8795840262 3.319065e-01 0.338401544</span></span>
<span><span class="co">#&gt; clsC2:trtB   0.0387066712 2.726210e-01 0.9114719020 4.580478e-01 0.548818677</span></span>
<span><span class="co">#&gt; clsC3:trtB   0.1322220329 1.338870e-01 0.7144983040 3.745743e-01 0.298129310</span></span>
<span><span class="co">#&gt; clsC4:trtB   0.7442524470 9.307711e-02 0.6485383571 5.182577e-01 0.080156444</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>We use a simulated multi-sample multi-cell-type scRNA-seq dataset to
illustrate how to utilize FLASHMM to perform single-cell differential
expression analysis. We are interested in identifying the genes
differentially expressed between two treatments (conditions) within a
cell type.</p>
<div class="section level3">
<h3 id="simulating-multi-sample-multi-cell-type-scrna-seq-dataset">Simulating multi-sample multi-cell-type scRNA-seq dataset<a class="anchor" aria-label="anchor" href="#simulating-multi-sample-multi-cell-type-scrna-seq-dataset"></a>
</h3>
<p>We generate a multi-sample multi-cell-type scRNA-seq dataset using a
reference dataset by simuRNAseq function in FLASHMM package. We use PBMC
10X droplet-based scRNA-seq dataset <span class="citation">(<a href="#ref-PBMC2018">Kang et al. 2018</a>)</span> as the reference
dataset, which contains count data and metadata. The metadata should
include 3 columns: individuals (subjects or samples), cell types
(clusters), and treatments (conditions) which are named as ‘sam’, ‘cls’,
and ‘trt’, respectively.</p>
<p>Note that the simuRNAseq function can also generate the scRNA-seq
dataset without a reference dataset. In this case, the following step
for preparing the reference dataset can be skipped.</p>
<div class="section level4">
<h4 id="preparing-reference-dataset">Preparing reference dataset<a class="anchor" aria-label="anchor" href="#preparing-reference-dataset"></a>
</h4>
<p>To load the PBMC dataset, we need ExperimentHub package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in get(paste0(generic, ".", class), envir = get_method_env()) : </span></span>
<span><span class="co">#&gt;   object 'type_sum.accel' not found</span></span>
<span></span>
<span><span class="co">##Load data.</span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#query(eh, "Kang")</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH2259"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co">##Remove undetected genes.</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span>
<span><span class="co">##Remove cells with few or many detected genes (outliers).</span></span>
<span><span class="va">nGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">bx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">nGenes</span><span class="op">)</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">nGenes</span> <span class="op">&gt;=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">bx</span><span class="op">$</span><span class="va">stats</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">nGenes</span> <span class="op">&lt;=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">bx</span><span class="op">$</span><span class="va">stats</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##Remove lowly expressed genes.</span></span>
<span><span class="co">##counts per cell (cpc) </span></span>
<span><span class="va">cpc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">cpc</span> <span class="op">&gt;</span> <span class="fl">0.005</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">##counts and metadata</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   ind stim cluster            cell multiplets</span></span>
<span><span class="co">#&gt; AAACATACAATGCC-1  107 ctrl       5     CD4 T cells    doublet</span></span>
<span><span class="co">#&gt; AAACATACATTTCC-1 1016 ctrl       9 CD14+ Monocytes    singlet</span></span>
<span><span class="co">#&gt; AAACATACCAGAAA-1 1256 ctrl       9 CD14+ Monocytes    singlet</span></span>
<span><span class="co">#&gt; AAACATACCAGCTA-1 1256 ctrl       9 CD14+ Monocytes    doublet</span></span>
<span><span class="co">#&gt; AAACATACCATGCA-1 1488 ctrl       3     CD4 T cells    singlet</span></span>
<span><span class="co">#&gt; AAACATACCTCGCT-1 1256 ctrl       9 CD14+ Monocytes    singlet</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  7483 28721</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">eh</span>, <span class="va">sce</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="generating-dataset">Generating dataset<a class="anchor" aria-label="anchor" href="#generating-dataset"></a>
</h4>
<p>We use the reference dataset to generate a scRNA-seq dataset by
simuRNAseq function. First we need to specify which columns represent
samples (individuals), cell clusters (types), and treatments
(experimental conditions) in the reference metadata by ‘sam’, ‘cls’, and
‘trt’, respectively. We also specify the numbers of genes, cells and
differentially expressed (DE) genes to be generated. We use default
settings for other arguments in simuRNAseq function. The generated
dataset contains count data, metadata, and the DE genes specific to a
cell cluster. The DE genes can be considered as the positive controls
and the others the negative controls. Without confusion, DE can
represent either ‘differentially expressed’ or ‘differential
expression’.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##Specify which columns represent samples, treatments, and cell-types.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="op">==</span> <span class="st">"ind"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"sam"</span>  <span class="co">#samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="op">==</span> <span class="st">"stim"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"trt"</span> <span class="co">#treatments</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="op">==</span> <span class="st">"cell"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"cls"</span> <span class="co">#cell-types</span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="va">coldata</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sam"</span>, <span class="st">"trt"</span>, <span class="st">"cls"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   sam  trt             cls</span></span>
<span><span class="co">#&gt; AAACATACAATGCC-1  107 ctrl     CD4 T cells</span></span>
<span><span class="co">#&gt; AAACATACATTTCC-1 1016 ctrl CD14+ Monocytes</span></span>
<span><span class="co">#&gt; AAACATACCAGAAA-1 1256 ctrl CD14+ Monocytes</span></span>
<span><span class="co">#&gt; AAACATACCAGCTA-1 1256 ctrl CD14+ Monocytes</span></span>
<span><span class="co">#&gt; AAACATACCATGCA-1 1488 ctrl     CD4 T cells</span></span>
<span><span class="co">#&gt; AAACATACCTCGCT-1 1256 ctrl CD14+ Monocytes</span></span>
<span><span class="co">##</span></span>
<span></span>
<span><span class="co">##Generate the dataset by simuRNAseq function.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2412</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simuRNAseq.html">simuRNAseq</a></span><span class="op">(</span><span class="va">counts</span>, nGenes <span class="op">=</span> <span class="fl">500</span>, nCells <span class="op">=</span> <span class="fl">120000</span>, metadata <span class="op">=</span> <span class="va">coldata</span>, </span>
<span>                  nsam <span class="op">=</span> <span class="fl">25</span>, ncls <span class="op">=</span> <span class="fl">10</span>, ntrt <span class="op">=</span> <span class="fl">2</span>, nDEgenes <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                  minbeta <span class="op">=</span> <span class="fl">0.5</span>, maxbeta <span class="op">=</span> <span class="fl">1</span>, var.randomeffects <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 5</span></span>
<span><span class="co">#&gt;  $ ref.mean.dispersion:'data.frame': 7483 obs. of  2 variables:</span></span>
<span><span class="co">#&gt;   ..$ mu        : num [1:7483] 0.0718 0.2008 19.708 0.0725 0.0755 ...</span></span>
<span><span class="co">#&gt;   ..$ dispersion: num [1:7483] 0.5183 0.0814 0.1902 0.0963 0.0483 ...</span></span>
<span><span class="co">#&gt;  $ metadata           :'data.frame': 120000 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ sam    : int [1:120000] 1488 1256 1488 1256 1015 1016 1016 1256 1256 1244 ...</span></span>
<span><span class="co">#&gt;   ..$ trt    : Factor w/ 2 levels "ctrl","stim": 2 1 1 1 1 2 1 2 2 2 ...</span></span>
<span><span class="co">#&gt;   ..$ cls    : Factor w/ 8 levels "B cells","CD14+ Monocytes",..: 3 2 3 1 4 6 4 3 2 3 ...</span></span>
<span><span class="co">#&gt;   ..$ libsize: num [1:120000] 1170 745 886 576 962 ...</span></span>
<span><span class="co">#&gt;  $ counts             : num [1:500, 1:120000] 0 0 1 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:500] "EPC2" "TNFRSF18" "TMEM87A" "SIRT6" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:120000] "GAGCATACGGAAGC-1" "CATGCCACAACTGC-1" "AGTGAAGAAGCGTT-1" "CTAACGGAGACACT-1" ...</span></span>
<span><span class="co">#&gt;  $ DEgenes            :'data.frame': 20 obs. of  3 variables:</span></span>
<span><span class="co">#&gt;   ..$ gene   : chr [1:20] "LNPEP" "C22orf39_ENSG00000242259" "MED29" "RANGAP1" ...</span></span>
<span><span class="co">#&gt;   ..$ beta   : num [1:20] 0.799 0.821 -0.911 -0.649 -0.896 ...</span></span>
<span><span class="co">#&gt;   ..$ cluster: chr [1:20] "B cells" "B cells" "CD14+ Monocytes" "CD4 T cells" ...</span></span>
<span><span class="co">#&gt;  $ treatment          : chr "stim"</span></span>
<span></span>
<span><span class="co">##Remove the reference dataset that is no longer needed in the following analysis.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">counts</span>, <span class="va">coldata</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="de-analysis-of-the-simulated-scrna-seq-data">DE analysis of the simulated scRNA-seq data<a class="anchor" aria-label="anchor" href="#de-analysis-of-the-simulated-scrna-seq-data"></a>
</h3>
<p>We perform differential expression analysis of the simulated dataset
using FLASHMM package. We are to identify the significantly
differentially expressed genes between two treatments in a cell-type.
The analyses involve following steps: LMM design, LMM fitting, and
exploring LMM fit and the DE genes.</p>
<p><strong>LMM design</strong>: construct design matrices for fixed and
random effects described as @ref(eq:lmm), and compute the gene
expression profile. The gene expression is taken as log-transformed
count matrix,
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Y</mi><mo>=</mo><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>+</mo><mtext mathvariant="normal">counts</mtext><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo></mrow><annotation encoding="application/x-tex">Y = \log(1+\text{counts}),</annotation></semantics></math>
in which each row corresponds to the expression profile for a gene. The
design matrix for the fixed effects can be created by the function
model.matrix, as follows
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mo>=</mo><mi>m</mi><mi>o</mi><mi>d</mi><mi>e</mi><mi>l</mi><mi>.</mi><mi>m</mi><mi>a</mi><mi>t</mi><mi>r</mi><mi>i</mi><mi>x</mi><mrow><mo stretchy="true" form="prefix">(</mo><mo>∼</mo><mn>0</mn><mo>+</mo><mi>l</mi><mi>o</mi><mi>g</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>l</mi><mi>i</mi><mi>b</mi><mi>r</mi><mi>a</mi><mi>r</mi><mi>y</mi><mi>.</mi><mi>s</mi><mi>i</mi><mi>z</mi><mi>e</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>.</mi><mi>t</mi><mi>y</mi><mi>p</mi><mi>e</mi><mo>+</mo><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>.</mi><mi>t</mi><mi>y</mi><mi>p</mi><mi>e</mi><mo>:</mo><mi>t</mi><mi>r</mi><mi>e</mi><mi>a</mi><mi>t</mi><mi>m</mi><mi>e</mi><mi>n</mi><mi>t</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo></mrow><annotation encoding="application/x-tex">
X = model.matrix(\sim 0 + log(library.size) + cell.type + cell.type:treatment),
</annotation></semantics></math> where the interaction term
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>.</mi><mi>t</mi><mi>y</mi><mi>p</mi><mi>e</mi><mo>:</mo><mi>t</mi><mi>r</mi><mi>e</mi><mi>a</mi><mi>t</mi><mi>m</mi><mi>e</mi><mi>n</mi><mi>t</mi></mrow><annotation encoding="application/x-tex">cell.type:treatment</annotation></semantics></math>
represents the treatment effect in a specific cell-type.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>i</mi><mi>b</mi><mi>r</mi><mi>a</mi><mi>r</mi><mi>y</mi><mi>.</mi><mi>s</mi><mi>i</mi><mi>z</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">library.size</annotation></semantics></math>
is defined as the total sum of counts across all genes for each cell, a
normalization factor of the scRNA-seq counts. We consider the subjects
(samples) as random effects that reflect the intra-subject correlation
and inter-subject variability. The design matrix for the random effects
is given by
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Z</mi><mo>=</mo><mi>m</mi><mi>o</mi><mi>d</mi><mi>e</mi><mi>l</mi><mi>.</mi><mi>m</mi><mi>a</mi><mi>t</mi><mi>r</mi><mi>i</mi><mi>x</mi><mrow><mo stretchy="true" form="prefix">(</mo><mo>∼</mo><mn>0</mn><mo>+</mo><mi>s</mi><mi>u</mi><mi>b</mi><mi>j</mi><mi>e</mi><mi>c</mi><mi>t</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>.</mi></mrow><annotation encoding="application/x-tex">
Z = model.matrix(\sim 0 + subject).
</annotation></semantics></math></p>
<p><strong>LMM fitting</strong>: We use either lmm or lmmfit function to
fit LMMs for all genes. With the cell-level data matrices
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Z</mi><annotation encoding="application/x-tex">Z</annotation></semantics></math>,
the LMMs can be fit by
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>m</mi><mi>m</mi><mi>f</mi><mi>i</mi><mi>t</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Y</mi><mo>,</mo><mi>X</mi><mo>,</mo><mi>Z</mi><mo>,</mo><mi>d</mi><mo>=</mo><mi>d</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo></mrow><annotation encoding="application/x-tex">lmmfit(Y, X, Z, d = d),</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>d</mi><annotation encoding="application/x-tex">d</annotation></semantics></math>
is the number of random-effects. For
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
components of random-effects,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>q</mi><mn>1</mn></msub><mo>,</mo><mi>…</mi><mo>,</mo><msub><mi>q</mi><mi>K</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">d = (q_1, \ldots, q_K)</annotation></semantics></math>,
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>q</mi><mi>k</mi></msub><annotation encoding="application/x-tex">q_k</annotation></semantics></math>
is the number of columns of the design matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Z</mi><mi>k</mi></msub><annotation encoding="application/x-tex">Z_k</annotation></semantics></math>
for the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>-th
random-effect component. For a large-scale data, the lmmfit function has
a problem of computer memory limit. In this case, it is recommended to
pre-compute the summary-level data:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mi>X</mi></mrow><annotation encoding="application/x-tex">XX</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mi>Y</mi></mrow><annotation encoding="application/x-tex">XY</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Z</mi><mi>X</mi></mrow><annotation encoding="application/x-tex">ZX</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Z</mi><mi>Y</mi></mrow><annotation encoding="application/x-tex">ZY</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Z</mi><mi>Z</mi></mrow><annotation encoding="application/x-tex">ZZ</annotation></semantics></math>,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>n</mi><mi>o</mi><mi>r</mi><mi>m</mi></mrow></msub><annotation encoding="application/x-tex">Y_{norm}</annotation></semantics></math>,
defined in @ref(eq:sdata), and then use the lmm function to fit the LMMs
as follows:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>m</mi><mi>m</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>X</mi><mi>X</mi><mo>,</mo><mi>X</mi><mi>Y</mi><mo>,</mo><mi>Z</mi><mi>X</mi><mo>,</mo><mi>Z</mi><mi>Y</mi><mo>,</mo><mi>Z</mi><mi>Z</mi><mo>,</mo><mi>Y</mi><mi>n</mi><mi>o</mi><mi>r</mi><mi>m</mi><mo>=</mo><mi>Y</mi><mi>n</mi><mi>o</mi><mi>r</mi><mi>m</mi><mo>,</mo><mi>n</mi><mo>=</mo><mi>n</mi><mo>,</mo><mi>d</mi><mo>=</mo><mi>d</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>.</mi></mrow><annotation encoding="application/x-tex">lmm(XX, XY, ZX, ZY, ZZ, Ynorm = Ynorm, n = n, d = d).</annotation></semantics></math>
The summary-level data doesn’t depend on the sample size
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>.
This makes lmm memory-efficient. Both lmm and lmmfit functions also
perform hypothesis testing on the fixed effects. We can use lmmtest
function to further perform statistical test on the contrasts of fixed
effects.</p>
<p>LMM fitting returns a list of estimates of LMM parameters
(coefficients and variance components), standard errors of the
estimates, covariance matrix of the coefficients (fixed effects), and
t-values and p-values for hypothesis testing on the coefficients, for
each gene. The LMM fitting also returns the numbers of iterations and
the first partial derivatives of log likelihood at the termination of
iterations for each gene.</p>
<p><strong>Exploring LMM fit and the DE genes</strong>: We check if the
LMM fitting is convergent, and perform hypothesis testing on the
variance components of random effects. Then we identify the DE genes
based on hypothesis testing p-values for the coefficients (fixed
effects). If the absolute first partial derivatives of log likelihood
are all less than the convergence tolerance, the LMM fitting converges,
otherwise it doesn’t converge. The genes for which LMM fitting doesn’t
converge should be excluded in the subsequent analysis because the
estimated coefficients for these genes are not reliable. The DE genes
can be identified by adjusting p-values obtained by false discovery rate
(FDR) or family-wise error rate (Bonferroni correction). We might
exclude the genes with small coefficients (effect size or
log-fold-change).</p>
<div class="section level4">
<h4 id="lmm-design">LMM design<a class="anchor" aria-label="anchor" href="#lmm-design"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##Gene expression matrix, Y = log2(1 + counts)</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">dat</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">dat</span><span class="op">$</span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="cn">NULL</span> <span class="co">#Remove the counts.</span></span>
<span></span>
<span><span class="co">##Design matrix for fixed effects</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">libsize</span><span class="op">)</span> <span class="op">+</span> <span class="va">cls</span> <span class="op">+</span> <span class="va">cls</span><span class="op">:</span><span class="va">trt</span>, data <span class="op">=</span> <span class="va">dat</span><span class="op">$</span><span class="va">meta</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\+"</span>, <span class="st">"p"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">" "</span>, <span class="st">"_"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##Design matrix for random effects</span></span>
<span><span class="va">Z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">sam</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">dat</span><span class="op">$</span><span class="va">meta</span><span class="op">)</span></span>
<span><span class="co">##Dimension of random effects</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="lmm-fitting">LMM fitting<a class="anchor" aria-label="anchor" href="#lmm-fitting"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  </span>
<span><span class="co">##(1) Fit LMM by cell-level data.</span></span>
<span><span class="va">max.iter</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">epsilon</span> <span class="op">&lt;-</span> <span class="fl">1e-5</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lmmfit.html">lmmfit</a></span><span class="op">(</span><span class="va">Y</span>, <span class="va">X</span>, <span class="va">Z</span>, d <span class="op">=</span> <span class="va">d</span>, max.iter <span class="op">=</span> <span class="va">max.iter</span>, epsilon <span class="op">=</span> <span class="va">epsilon</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 12</span></span>
<span><span class="co">#&gt;  $ method  : chr "REML-FS"</span></span>
<span><span class="co">#&gt;  $ dlogL   : num [1:2, 1:500] 4.27e-07 -9.31e-10 6.08e-06 -8.50e-09 4.85e-06 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : NULL</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:500] "dl" "dl" "dl" "dl" ...</span></span>
<span><span class="co">#&gt;  $ niter   : num [1:500] 7 6 7 7 14 12 12 7 8 6 ...</span></span>
<span><span class="co">#&gt;  $ coef    : num [1:17, 1:500] 0.0171 -0.1112 -0.1096 -0.109 -0.1055 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:17] "log(libsize)" "clsB_cells" "clsCD14p_Monocytes" "clsCD4_T_cells" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:500] "EPC2" "TNFRSF18" "TMEM87A" "SIRT6" ...</span></span>
<span><span class="co">#&gt;  $ se      : num [1:17, 1:500] 0.000991 0.007452 0.007863 0.007287 0.007398 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:17] "log(libsize)" "clsB_cells" "clsCD14p_Monocytes" "clsCD4_T_cells" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:500] "EPC2" "TNFRSF18" "TMEM87A" "SIRT6" ...</span></span>
<span><span class="co">#&gt;  $ t       : num [1:17, 1:500] 17.3 -14.9 -13.9 -15 -14.3 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:17] "log(libsize)" "clsB_cells" "clsCD14p_Monocytes" "clsCD4_T_cells" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:500] "EPC2" "TNFRSF18" "TMEM87A" "SIRT6" ...</span></span>
<span><span class="co">#&gt;  $ p       : num [1:17, 1:500] 7.29e-67 2.88e-50 3.82e-44 1.60e-50 4.11e-46 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:17] "log(libsize)" "clsB_cells" "clsCD14p_Monocytes" "clsCD4_T_cells" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:500] "EPC2" "TNFRSF18" "TMEM87A" "SIRT6" ...</span></span>
<span><span class="co">#&gt;  $ cov     : num [1:17, 1:17, 1:500] 9.81e-07 -6.95e-06 -7.49e-06 -6.93e-06 -6.87e-06 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 3</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:17] "log(libsize)" "clsB_cells" "clsCD14p_Monocytes" "clsCD4_T_cells" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:17] "log(libsize)" "clsB_cells" "clsCD14p_Monocytes" "clsCD4_T_cells" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:500] "EPC2" "TNFRSF18" "TMEM87A" "SIRT6" ...</span></span>
<span><span class="co">#&gt;  $ df      : int 119983</span></span>
<span><span class="co">#&gt;  $ theta   : num [1:2, 1:500] 2.68e-05 1.80e-02 1.12e-04 7.88e-02 1.91e-04 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:2] "var1" "var0"</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:500] "EPC2" "TNFRSF18" "TMEM87A" "SIRT6" ...</span></span>
<span><span class="co">#&gt;  $ se.theta: num [1:2, 1:500] 1.52e-05 7.34e-05 6.36e-05 3.22e-04 1.06e-04 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:2] "var1" "var0"</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:500] "EPC2" "TNFRSF18" "TMEM87A" "SIRT6" ...</span></span>
<span><span class="co">#&gt;  $ RE      : NULL</span></span>
<span><span class="co">##</span></span>
<span></span>
<span><span class="co">##(2) Fit LMM by summary-level data.</span></span>
<span><span class="co">##- Compute the summary-level data.</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">XX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">X</span></span>
<span><span class="va">XY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">Y</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">ZX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">X</span></span>
<span><span class="va">ZY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">Y</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">Z</span><span class="op">)</span></span>
<span><span class="va">ZZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">Z</span></span>
<span><span class="va">Ynorm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">Y</span><span class="op">*</span><span class="va">Y</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span>, <span class="va">Z</span><span class="op">)</span> <span class="co">#release the memory.</span></span>
<span></span>
<span><span class="co">##- Fit LMM.</span></span>
<span><span class="va">fitss</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lmm.html">lmm</a></span><span class="op">(</span><span class="va">XX</span>, <span class="va">XY</span>, <span class="va">ZX</span>, <span class="va">ZY</span>, <span class="va">ZZ</span>, Ynorm <span class="op">=</span> <span class="va">Ynorm</span>, n <span class="op">=</span> <span class="va">n</span>, d <span class="op">=</span> <span class="va">d</span>, </span>
<span>             max.iter <span class="op">=</span> <span class="va">max.iter</span>, epsilon <span class="op">=</span> <span class="va">epsilon</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">fitss</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">fitss</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##Test the treatment effect within all cell-types.</span></span>
<span><span class="co">##- Construct a contrast by summing the treatment effects across cell-types.</span></span>
<span><span class="va">contrast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="st">"trt"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">coef</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">contrast</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">":"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">coef</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="co">##- Test the contrast.</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lmmtest.html">lmmtest</a></span><span class="op">(</span><span class="va">fit</span>, contrast <span class="op">=</span> <span class="va">contrast</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span></span>
<span><span class="co">#&gt;               trt_t     trt_p</span></span>
<span><span class="co">#&gt; EPC2     -1.4955064 0.1347850</span></span>
<span><span class="co">#&gt; TNFRSF18  1.3729485 0.1697709</span></span>
<span><span class="co">#&gt; TMEM87A  -0.1572237 0.8750688</span></span>
<span><span class="co">#&gt; SIRT6    -0.7064178 0.4799297</span></span>
<span><span class="co">#&gt; IL12RB1  -1.6251549 0.1041322</span></span>
<span><span class="co">#&gt; RBFA      0.6404263 0.5218967</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="exploring-lmm-fit-and-the-de-genes">Exploring LMM fit and the DE genes<a class="anchor" aria-label="anchor" href="#exploring-lmm-fit-and-the-de-genes"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">##(1) Check which LMM fittings converge.</span></span>
<span><span class="va">cvg</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">dlogL</span><span class="op">)</span>, <span class="fl">2</span>, <span class="va">max</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">epsilon</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">cvg</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] 500</span></span>
<span><span class="co">##</span></span>
<span></span>
<span><span class="co">##(2) Hypothesis testing for variance components:</span></span>
<span><span class="co">##    H0, theta &lt;/= 0 vs H1, theta &gt; 0.</span></span>
<span><span class="va">z</span> <span class="op">&lt;-</span> <span class="va">fit</span><span class="op">$</span><span class="va">theta</span><span class="op">[</span><span class="st">"var1"</span>, <span class="op">]</span><span class="op">/</span><span class="va">fit</span><span class="op">$</span><span class="va">se.theta</span><span class="op">[</span><span class="st">"var1"</span>, <span class="op">]</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">pnorm</a></span><span class="op">(</span><span class="va">z</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">p</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 461</span></span>
<span><span class="co">##</span></span>
<span></span>
<span><span class="co">##(3) The DE genes specific to a cell-type</span></span>
<span><span class="co">##Coefficients and p-values for the genes specific to a cell-type.</span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">":"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">coef</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">beta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="va">index</span>, <span class="va">cvg</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">p</span><span class="op">[</span><span class="va">index</span>, <span class="va">cvg</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##Adjust p-values by FDR.</span></span>
<span><span class="va">padj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"fdr"</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##DE genes specific to a cell cluster with FDR &lt; 0.05.</span></span>
<span><span class="va">degenes</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">i</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">padj</span><span class="op">[</span>, <span class="va">j</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span>  <span class="va">degenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">degenes</span>, <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, cluster <span class="op">=</span> <span class="va">j</span>, coef <span class="op">=</span> <span class="va">beta</span><span class="op">[</span><span class="va">i</span>, <span class="va">j</span><span class="op">]</span>, p <span class="op">=</span> <span class="va">p</span><span class="op">[</span><span class="va">i</span>, <span class="va">j</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">degenes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">degenes</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 31  4</span></span>
<span><span class="va">degenes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;                        gene cluster        coef            p</span></span>
<span><span class="co">#&gt; 1                     LNPEP       1 0.038532290 2.205190e-20</span></span>
<span><span class="co">#&gt; 2  C22orf39_ENSG00000242259       1 0.061134158 3.618583e-37</span></span>
<span><span class="co">#&gt; 3                     EIF3G       2 0.024793616 5.835875e-05</span></span>
<span><span class="co">#&gt; 4    FAM58A_ENSG00000147382       2 0.013428790 7.854205e-07</span></span>
<span><span class="co">#&gt; 5                       CD7       2 0.030955080 2.766240e-04</span></span>
<span><span class="co">#&gt; 6                      TMPO       2 0.009871346 1.898710e-04</span></span>
<span><span class="co">#&gt; 7                      GPS2       2 0.008835609 4.649781e-05</span></span>
<span><span class="co">#&gt; 8                    FAM96A       2 0.014721218 3.116602e-04</span></span>
<span><span class="co">#&gt; 9                    CSTF2T       2 0.011218177 3.093001e-04</span></span>
<span><span class="co">#&gt; 10                    EIF3M       2 0.024587165 6.576295e-05</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Monterey 12.7.6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Toronto</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] muscData_1.18.0             SingleCellExperiment_1.26.0</span></span>
<span><span class="co">#&gt;  [3] SummarizedExperiment_1.34.0 Biobase_2.64.0             </span></span>
<span><span class="co">#&gt;  [5] GenomicRanges_1.56.2        GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">#&gt;  [7] IRanges_2.38.1              S4Vectors_0.42.1           </span></span>
<span><span class="co">#&gt;  [9] MatrixGenerics_1.16.0       matrixStats_1.4.1          </span></span>
<span><span class="co">#&gt; [11] ExperimentHub_2.12.0        AnnotationHub_3.12.0       </span></span>
<span><span class="co">#&gt; [13] BiocFileCache_2.12.0        dbplyr_2.5.0               </span></span>
<span><span class="co">#&gt; [15] BiocGenerics_0.50.0         FLASHMM_0.1.0              </span></span>
<span><span class="co">#&gt; [17] devtools_2.4.5              usethis_3.1.0              </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1        dplyr_1.1.4             blob_1.2.4             </span></span>
<span><span class="co">#&gt;  [4] filelock_1.0.3          Biostrings_2.72.1       fastmap_1.2.0          </span></span>
<span><span class="co">#&gt;  [7] promises_1.3.2          digest_0.6.37           mime_0.12              </span></span>
<span><span class="co">#&gt; [10] lifecycle_1.0.4         ellipsis_0.3.2          processx_3.8.4         </span></span>
<span><span class="co">#&gt; [13] KEGGREST_1.44.1         RSQLite_2.3.9           magrittr_2.0.3         </span></span>
<span><span class="co">#&gt; [16] compiler_4.4.2          rlang_1.1.4             sass_0.4.9             </span></span>
<span><span class="co">#&gt; [19] tools_4.4.2             yaml_2.3.10             knitr_1.49             </span></span>
<span><span class="co">#&gt; [22] S4Arrays_1.4.1          htmlwidgets_1.6.4       bit_4.5.0.1            </span></span>
<span><span class="co">#&gt; [25] pkgbuild_1.4.5          curl_6.0.1              DelayedArray_0.30.1    </span></span>
<span><span class="co">#&gt; [28] abind_1.4-8             pkgload_1.4.0           miniUI_0.1.1.1         </span></span>
<span><span class="co">#&gt; [31] withr_3.0.2             purrr_1.0.2             desc_1.4.3             </span></span>
<span><span class="co">#&gt; [34] grid_4.4.2              urlchecker_1.0.1        profvis_0.4.0          </span></span>
<span><span class="co">#&gt; [37] xtable_1.8-4            MASS_7.3-64             cli_3.6.3              </span></span>
<span><span class="co">#&gt; [40] rmarkdown_2.29          crayon_1.5.3            ragg_1.3.3             </span></span>
<span><span class="co">#&gt; [43] generics_0.1.3          remotes_2.5.0           rstudioapi_0.17.1      </span></span>
<span><span class="co">#&gt; [46] httr_1.4.7              sessioninfo_1.2.2       DBI_1.2.3              </span></span>
<span><span class="co">#&gt; [49] cachem_1.1.0            zlibbioc_1.50.0         AnnotationDbi_1.66.0   </span></span>
<span><span class="co">#&gt; [52] BiocManager_1.30.25     XVector_0.44.0          vctrs_0.6.5            </span></span>
<span><span class="co">#&gt; [55] Matrix_1.7-1            jsonlite_1.8.9          bookdown_0.41          </span></span>
<span><span class="co">#&gt; [58] callr_3.7.6             bit64_4.5.2             systemfonts_1.1.0      </span></span>
<span><span class="co">#&gt; [61] jquerylib_0.1.4         glue_1.8.0              pkgdown_2.1.1          </span></span>
<span><span class="co">#&gt; [64] ps_1.8.1                BiocVersion_3.19.1      later_1.4.1            </span></span>
<span><span class="co">#&gt; [67] UCSC.utils_1.0.0        tibble_3.2.1            pillar_1.10.0          </span></span>
<span><span class="co">#&gt; [70] rappdirs_0.3.3          htmltools_0.5.8.1       GenomeInfoDbData_1.2.12</span></span>
<span><span class="co">#&gt; [73] R6_2.5.1                textshaping_0.4.1       evaluate_1.0.1         </span></span>
<span><span class="co">#&gt; [76] shiny_1.10.0            lattice_0.22-6          png_0.1-8              </span></span>
<span><span class="co">#&gt; [79] memoise_2.0.1           httpuv_1.6.15           bslib_0.8.0            </span></span>
<span><span class="co">#&gt; [82] Rcpp_1.0.13-1           SparseArray_1.4.8       xfun_0.49              </span></span>
<span><span class="co">#&gt; [85] fs_1.6.5                pkgconfig_2.0.3</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="remarks">Remarks<a class="anchor" aria-label="anchor" href="#remarks"></a>
</h2>
<p>Building design matrices for fixed and random effects is a key step
for LMM-based DE analysis. Including library size, a normalization
factor for scRNA-seq, as a fixed effect can help reduce p-value
inflation. If needed, we can add principal components (PCs) as fixed
effects to further remove the unknown batch effect.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-HartleyRao1967" class="csl-entry">
Hartley, H. O., and J. N. K. Rao. 1967. <span>“Maximum-Likelihood
Estimation for the Mixed Analysis of Variance Model.”</span>
<em>Biometrika</em> 54 (1/2): 93–108. <a href="http://www.jstor.org/stable/2333854" class="external-link">http://www.jstor.org/stable/2333854</a>.
</div>
<div id="ref-PBMC2018" class="csl-entry">
Kang, Hyun Min, Meena Subramaniam, Sasha Targ, Michelle Nguyen, Lenka
Maliskova, Elizabeth McCarthy, Eunice Wan, et al. 2018.
<span>“Multiplexed Droplet Single-Cell RNA-Sequencing Using Natural
Genetic Variation.”</span> <em>Nature Biotechnology</em> 36 (89–94). <a href="https://doi.org/10.1038/nbt.4042" class="external-link">https://doi.org/10.1038/nbt.4042</a>.
</div>
<div id="ref-PattersonThompson1971" class="csl-entry">
Patterson, H. D., and R. Thompson. 1971. <span>“Recovery of Inter-Block
Information When Block Sizes Are Unequal.”</span> <em>Biometrika</em> 58
(3): 545–54. <a href="https://doi.org/10.2307/2334389" class="external-link">https://doi.org/10.2307/2334389</a>.
</div>
<div id="ref-Searle2006" class="csl-entry">
Searle, Shayle R., George Casella, and Charles E. McCulloch. 2006.
<em>Variance Components</em>. New Jersey: John Wiley &amp; Sons, Inc.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Changjiang Xu, Gary Bader.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
