<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Fitting Linear Mixed-effects Models — lmmfit • FLASHMM</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Fitting Linear Mixed-effects Models — lmmfit"><meta name="description" content="lmmfit, a wrapper function of lmm, fits linear mixed-effects models (LMM) by sample-level data. The LMM parameters are estimated by restricted maximum likelihood (REML) with Fisher scoring (FS) gradient descent algorithm."><meta property="og:description" content="lmmfit, a wrapper function of lmm, fits linear mixed-effects models (LMM) by sample-level data. The LMM parameters are estimated by restricted maximum likelihood (REML) with Fisher scoring (FS) gradient descent algorithm."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLASHMM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/FLASHMM-vignette.html">Single-cell differential expression analysis with FLASHMM</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/BaderLab/FLASHMM/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Fitting Linear Mixed-effects Models</h1>
      <small class="dont-index">Source: <a href="https://github.com/BaderLab/FLASHMM/blob/HEAD/R/lmmfit.R"><code>R/lmmfit.R</code></a></small>
      <div class="d-none name"><code>lmmfit.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>lmmfit, a wrapper function of lmm, fits linear mixed-effects models (LMM) by sample-level data. The LMM parameters are estimated by restricted maximum likelihood (REML) with Fisher scoring (FS) gradient descent algorithm.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">lmmfit</span><span class="op">(</span></span>
<span>  <span class="va">Y</span>,</span>
<span>  <span class="va">X</span>,</span>
<span>  <span class="va">Z</span>,</span>
<span>  <span class="va">d</span>,</span>
<span>  theta0 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"REML-FS"</span>,</span>
<span>  max.iter <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  epsilon <span class="op">=</span> <span class="fl">1e-05</span>,</span>
<span>  output.cov <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  output.RE <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-y">Y<a class="anchor" aria-label="anchor" href="#arg-y"></a></dt>
<dd><p>A features-by-samples matrix of responses (genes-by-cells matrix of gene expressions for scRNA-seq).</p></dd>


<dt id="arg-x">X<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A design matrix for fixed effects, with rows corresponding to the columns of Y.</p></dd>


<dt id="arg-z">Z<a class="anchor" aria-label="anchor" href="#arg-z"></a></dt>
<dd><p>A design matrix for random effects, with rows corresponding to the columns of Y. Z = [Z1, ..., Zk], and Zi, i=1,...,k, is the design matrix for the i-th type random factor.</p></dd>


<dt id="arg-d">d<a class="anchor" aria-label="anchor" href="#arg-d"></a></dt>
<dd><p>A vector of (m1,...,mk), mi = ncol(Zi), number of columns in Zi. m1 + ... + mk = ncol(Z), number of columns in Z.</p></dd>


<dt id="arg-theta-">theta0<a class="anchor" aria-label="anchor" href="#arg-theta-"></a></dt>
<dd><p>A vector of initial values of the variance components, (s1, ...,sk, s_(k+1)), si = sigma_i^2, the variance component of the i-th type random effects. s_(k+1) = sigma^2, the variance component of model residual error.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>The REML with Fisher scoring (FS) iterative algorithm, REML-FS.</p></dd>


<dt id="arg-max-iter">max.iter<a class="anchor" aria-label="anchor" href="#arg-max-iter"></a></dt>
<dd><p>The maximal number of iterations for the iterative algorithm.</p></dd>


<dt id="arg-epsilon">epsilon<a class="anchor" aria-label="anchor" href="#arg-epsilon"></a></dt>
<dd><p>Positive convergence tolerance. If the absolute value of the first partial derivative of log likelihood is less than epsilon, the iterations converge.</p></dd>


<dt id="arg-output-cov">output.cov<a class="anchor" aria-label="anchor" href="#arg-output-cov"></a></dt>
<dd><p>If TRUE, output the covariance matrices for the estimated coefficients, which are needed for testing contrasts.</p></dd>


<dt id="arg-output-re">output.RE<a class="anchor" aria-label="anchor" href="#arg-output-re"></a></dt>
<dd><p>If TRUE, output the best linear unbiased prediction (BLUP) of the random effects.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the following components:</p>
<p>dlogL First partial derivatives of log-likelihoods for each feature (gene).</p>
<p>niter Nmbers of iterations for each feature (gene).</p>
<p>coef A matrix of estimated coefficients (fixed effects), each column corresponds to a feature (gene) and each row one covariate.</p>
<p>se A matrix of the standard errors of the estimated coefficients.</p>
<p>t A matrix of t-values for the fixed effects, equal to coef/se.</p>
<p>df Degrees of freedom.</p>
<p>p A matrix of two-sided p-values for the fixed effects.</p>
<p>cov A array of covariance matrices of the estimated coefficients (fixed effects).</p>
<p>theta A matrix of the estimated variance components, each column corresponds to a feature (gene) and each row one variance component. The last row is the variance component of the residual error.</p>
<p>se.theta Standard errors of the estimated theta.</p>
<p>RE A matrix of the best linear unbiased prediction (BLUP) of random effects.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#Generate data: X, Y, and Z.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">1e3</span></span></span>
<span class="r-in"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">10</span></span></span>
<span class="r-in"><span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">m</span><span class="op">*</span><span class="va">n</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">trt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="va">n</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">trt</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">20</span></span></span>
<span class="r-in"><span><span class="va">sam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="va">n</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sam</span><span class="op">[</span><span class="va">trt</span> <span class="op">==</span> <span class="st">"A"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample.int</a></span><span class="op">(</span><span class="va">q</span><span class="op">/</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">trt</span> <span class="op">==</span> <span class="st">"A"</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sam</span><span class="op">[</span><span class="va">trt</span> <span class="op">==</span> <span class="st">"B"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"B"</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample.int</a></span><span class="op">(</span><span class="va">q</span><span class="op">/</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">trt</span> <span class="op">==</span> <span class="st">"B"</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">Z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">sam</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Fitting LMM by the cell-level data</span></span></span>
<span class="r-in"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">lmmfit</span><span class="op">(</span><span class="va">Y</span>, <span class="va">X</span>, <span class="va">Z</span>, d <span class="op">=</span> <span class="va">d</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ method  : chr "REML-FS"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ dlogL   : num [1:2, 1:10] -1.34e-06 -8.32e-09 -1.02e-06 -1.38e-09 9.52e-07 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:10] "dl" "dl" "dl" "dl" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ niter   : num [1:10] 10 7 6 8 7 7 5 6 10 11</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ coef    : num [1:2, 1:10] 0.03745 0.01666 -0.04225 -0.02294 -0.00892 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:2] "trtA" "trtB"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:10] "Gene1" "Gene2" "Gene3" "Gene4" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ se      : num [1:2, 1:10] 0.0378 0.0376 0.0426 0.0424 0.0552 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:2] "trtA" "trtB"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:10] "Gene1" "Gene2" "Gene3" "Gene4" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ t       : num [1:2, 1:10] 0.992 0.443 -0.992 -0.541 -0.162 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:2] "trtA" "trtB"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:10] "Gene1" "Gene2" "Gene3" "Gene4" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ p       : num [1:2, 1:10] 0.322 0.658 0.321 0.589 0.872 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:2] "trtA" "trtB"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:10] "Gene1" "Gene2" "Gene3" "Gene4" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ cov     : num [1:2, 1:2, 1:10] 0.00143 0 0 0.00141 0.00181 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "dimnames")=List of 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:2] "trtA" "trtB"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:2] "trtA" "trtB"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:10] "Gene1" "Gene2" "Gene3" "Gene4" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ df      : int 998</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ theta   : num [1:2, 1:10] -0.00641 1.03645 -0.00132 0.97027 0.01133 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:2] "var1" "var0"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:10] "Gene1" "Gene2" "Gene3" "Gene4" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ se.theta: num [1:2, 1:10] 0.00478 0.04681 0.00606 0.04383 0.01013 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:2] "var1" "var0"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:10] "Gene1" "Gene2" "Gene3" "Gene4" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ RE      : NULL</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Hypothesis tests</span></span></span>
<span class="r-in"><span><span class="fu"><a href="lmmtest.html">lmmtest</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             trtA_t     trtB_t     trtA_p    trtB_p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene1   0.99151700  0.4430047 0.32167353 0.6578584</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene2  -0.99188018 -0.5407880 0.32149640 0.5887744</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene3  -0.16176056 -0.1851246 0.87152715 0.8531689</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene4  -0.56751196 -0.8348433 0.57049404 0.4040058</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene5  -0.97591263 -0.2341318 0.32934425 0.8149307</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene6  -0.05543855 -0.7219293 0.95580018 0.4705070</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene7  -1.17149564 -0.6587799 0.24167922 0.5101891</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene8   0.10299581  2.2227291 0.91798699 0.0264575</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene9   1.32183317  0.3175327 0.18652661 0.7509059</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene10 -2.00516281  0.2376670 0.04521529 0.8121882</span>
<span class="r-in"><span><span class="fu"><a href="lmmtest.html">lmmtest</a></span><span class="op">(</span><span class="va">fit</span>, index <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            trtB_t    trtB_p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene1   0.4430047 0.6578584</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene2  -0.5407880 0.5887744</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene3  -0.1851246 0.8531689</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene4  -0.8348433 0.4040058</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene5  -0.2341318 0.8149307</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene6  -0.7219293 0.4705070</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene7  -0.6587799 0.5101891</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene8   2.2227291 0.0264575</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene9   0.3175327 0.7509059</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene10  0.2376670 0.8121882</span>
<span class="r-in"><span><span class="fu"><a href="lmmtest.html">lmmtest</a></span><span class="op">(</span><span class="va">fit</span>, contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="st">"B-A"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              B-A_t     B-A_p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene1  -0.39018951 0.6964797</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene2   0.32121982 0.7481111</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene3  -0.01616384 0.9871069</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene4  -0.18729690 0.8514659</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene5   0.52649028 0.5986646</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene6  -0.47033528 0.6382183</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene7   0.36502924 0.7151669</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene8   1.49642518 0.1348591</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene9  -0.71274410 0.4761707</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene10  1.58872838 0.1124385</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Changjiang Xu, Gary Bader.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

