<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Simulating Multi-sample Multi-cell-type scRNA-seq Dataset based on Negative Binomial Distribution — simuRNAseq • FLASHMM</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulating Multi-sample Multi-cell-type scRNA-seq Dataset based on Negative Binomial Distribution — simuRNAseq"><meta name="description" content="simuRNAseq simulates scRNA-seq data with multiple subjects (samples), multiple clusters (cell-types) and two treatments (conditions) based on a negative binomial (NB) distribution using a reference data as background or control.
The reference data consisting of genes-by-cells counts matrix is used to estimate the NB dispersion and means for the genes to be simulated. The simulated genes are randomly selected from the reference data.
The NB dispersion are estimated by the method-of-moments estimate (MME).
The NB means for the background in the control are estimated by the sample mean. The NB means for the differentially expressed (DE) genes are given by the sample mean plus a log-fold change (logFC).
The simulated cells are randomly selected from the meta data that specifies subjects, cell-types and treatments for the cells.
The meta data consists of samples, clusters of cell types, and treatments, which can be generated either from reference data or randomly. If not provided, it will be randomly generated.
A random seed is recommended to be specified by set.seed before simulating data."><meta property="og:description" content="simuRNAseq simulates scRNA-seq data with multiple subjects (samples), multiple clusters (cell-types) and two treatments (conditions) based on a negative binomial (NB) distribution using a reference data as background or control.
The reference data consisting of genes-by-cells counts matrix is used to estimate the NB dispersion and means for the genes to be simulated. The simulated genes are randomly selected from the reference data.
The NB dispersion are estimated by the method-of-moments estimate (MME).
The NB means for the background in the control are estimated by the sample mean. The NB means for the differentially expressed (DE) genes are given by the sample mean plus a log-fold change (logFC).
The simulated cells are randomly selected from the meta data that specifies subjects, cell-types and treatments for the cells.
The meta data consists of samples, clusters of cell types, and treatments, which can be generated either from reference data or randomly. If not provided, it will be randomly generated.
A random seed is recommended to be specified by set.seed before simulating data."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLASHMM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/FLASHMM-vignette.html">Single-cell differential expression analysis with FLASHMM</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/BaderLab/FLASHMM/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Simulating Multi-sample Multi-cell-type scRNA-seq Dataset based on Negative Binomial Distribution</h1>
      <small class="dont-index">Source: <a href="https://github.com/BaderLab/FLASHMM/blob/HEAD/R/simuRNAseq.R"><code>R/simuRNAseq.R</code></a></small>
      <div class="d-none name"><code>simuRNAseq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>simuRNAseq simulates scRNA-seq data with multiple subjects (samples), multiple clusters (cell-types) and two treatments (conditions) based on a negative binomial (NB) distribution using a reference data as background or control.
The reference data consisting of genes-by-cells counts matrix is used to estimate the NB dispersion and means for the genes to be simulated. The simulated genes are randomly selected from the reference data.
The NB dispersion are estimated by the method-of-moments estimate (MME).
The NB means for the background in the control are estimated by the sample mean. The NB means for the differentially expressed (DE) genes are given by the sample mean plus a log-fold change (logFC).
The simulated cells are randomly selected from the meta data that specifies subjects, cell-types and treatments for the cells.
The meta data consists of samples, clusters of cell types, and treatments, which can be generated either from reference data or randomly. If not provided, it will be randomly generated.
A random seed is recommended to be specified by set.seed before simulating data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">simuRNAseq</span><span class="op">(</span></span>
<span>  <span class="va">counts</span>,</span>
<span>  nGenes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  nCells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  samples.nested <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  nsam <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  ncls <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  ntrt <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  trt <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nDEgenes <span class="op">=</span> <span class="va">ncls</span>,</span>
<span>  <span class="va">nDEgenesType</span>,</span>
<span>  pDEgenesType <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  adjust.library.size <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  direction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"both"</span>, <span class="st">"up"</span>, <span class="st">"down"</span><span class="op">)</span>,</span>
<span>  minbeta <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  maxbeta <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  var.randomeffects <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>A genes-by-cells matrix of reference counts. If missing, counts is generated by a negative binomial distribution.</p></dd>


<dt id="arg-ngenes">nGenes<a class="anchor" aria-label="anchor" href="#arg-ngenes"></a></dt>
<dd><p>Number of genes to be simulated.</p></dd>


<dt id="arg-ncells">nCells<a class="anchor" aria-label="anchor" href="#arg-ncells"></a></dt>
<dd><p>Number of cells to be simulated.</p></dd>


<dt id="arg-metadata">metadata<a class="anchor" aria-label="anchor" href="#arg-metadata"></a></dt>
<dd><p>The meta data consisting of 4 columns: sam (sample labels), cls (cluster lables of cell types), trt (treatments or conditions) and libsize (library size or total counts per cell), which is randomly generated if not provided.</p></dd>


<dt id="arg-samples-nested">samples.nested<a class="anchor" aria-label="anchor" href="#arg-samples-nested"></a></dt>
<dd><p>If TRUE, when metadata is not provided, each simulated subject (sample) belongs to only one condition (either treatment or control), that is, the subject is nested in a condition (treatment).</p></dd>


<dt id="arg-nsam">nsam<a class="anchor" aria-label="anchor" href="#arg-nsam"></a></dt>
<dd><p>Number of subjects (individuals).</p></dd>


<dt id="arg-ncls">ncls<a class="anchor" aria-label="anchor" href="#arg-ncls"></a></dt>
<dd><p>Number of clusters (cell-types).</p></dd>


<dt id="arg-ntrt">ntrt<a class="anchor" aria-label="anchor" href="#arg-ntrt"></a></dt>
<dd><p>Number of treatments (only one condition is treated).</p></dd>


<dt id="arg-trt">trt<a class="anchor" aria-label="anchor" href="#arg-trt"></a></dt>
<dd><p>Treatment, specifying which condition is treated.</p></dd>


<dt id="arg-ndegenes">nDEgenes<a class="anchor" aria-label="anchor" href="#arg-ndegenes"></a></dt>
<dd><p>Total number of DE genes.</p></dd>


<dt id="arg-ndegenestype">nDEgenesType<a class="anchor" aria-label="anchor" href="#arg-ndegenestype"></a></dt>
<dd><p>Number of DE genes specific to a cell type, named by cell cluster labels.</p></dd>


<dt id="arg-pdegenestype">pDEgenesType<a class="anchor" aria-label="anchor" href="#arg-pdegenestype"></a></dt>
<dd><p>Proportion of DE genes in a cell-type. Default NULL means equal proportion.</p></dd>


<dt id="arg-adjust-library-size">adjust.library.size<a class="anchor" aria-label="anchor" href="#arg-adjust-library-size"></a></dt>
<dd><p>If TRUE, adjust library sizes using the reference counts.</p></dd>


<dt id="arg-direction">direction<a class="anchor" aria-label="anchor" href="#arg-direction"></a></dt>
<dd><p>Specify if the DE genes are up- and/or down-regulated.</p></dd>


<dt id="arg-minbeta">minbeta<a class="anchor" aria-label="anchor" href="#arg-minbeta"></a></dt>
<dd><p>Lower bound of the DE gene logFC.</p></dd>


<dt id="arg-maxbeta">maxbeta<a class="anchor" aria-label="anchor" href="#arg-maxbeta"></a></dt>
<dd><p>Upper bound of the DE gene logFC. minbeta &lt; maxbeta. If direction = "both", minbeta*maxbeta &gt; 0. If direction = "up", minbeta &gt; 0. If direction = "down", maxbeta &lt; 0.</p></dd>


<dt id="arg-var-randomeffects">var.randomeffects<a class="anchor" aria-label="anchor" href="#arg-var-randomeffects"></a></dt>
<dd><p>Variance of random effects</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>ref.mean.dispersion: A data frame of the reference counts' means and dispersion,</p>
<p>metadata: Meta data consisting of 4 columns: sam (sample labels), cls (cluster lables of cell types), trt (two treatments/conditions) and libsize (library sizes).</p>
<p>counts: A genes-by-cells matrix of the simulated counts.</p>
<p>DEgenes: A data frame of DE genes consisting of 3 columns: gene, beta (effect), and cluster to which the gene is specific.</p>
<p>treatment: The condition treated.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#Simulate a multi-sample multi-cell-type scRNA-seq dataset.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2412</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">simuRNAseq</span><span class="op">(</span>nGenes <span class="op">=</span> <span class="fl">50</span>, nCells <span class="op">=</span> <span class="fl">1000</span>, nsam <span class="op">=</span> <span class="fl">25</span>, ncls <span class="op">=</span> <span class="fl">4</span>, ntrt <span class="op">=</span> <span class="fl">2</span>, nDEgenes <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Message: the condition B is treated.</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ ref.mean.dispersion:'data.frame':	50 obs. of  2 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ mu        : num [1:50] 1.077 0.992 2.815 1.945 2.06 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ dispersion: num [1:50] 1.602 2.989 0.954 2.937 1.914 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ metadata           :'data.frame':	1000 obs. of  4 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ sam    : chr [1:1000] "B1" "A6" "A2" "B8" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ cls    : chr [1:1000] "C4" "C3" "C1" "C1" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ trt    : chr [1:1000] "B" "A" "A" "B" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ libsize: num [1:1000] 117 75 101 80 123 113 125 95 77 95 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ counts             : num [1:50, 1:1000] 0 2 6 2 5 2 0 0 2 11 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..- attr(*, "dimnames")=List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:50] "Gene1" "Gene2" "Gene3" "Gene4" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ : chr [1:1000] "Cell1" "Cell2" "Cell3" "Cell4" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ DEgenes            :'data.frame':	6 obs. of  3 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ gene   : chr [1:6] "Gene45" "Gene46" "Gene47" "Gene48" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ beta   : num [1:6] 0.55 0.365 0.831 -0.948 -0.875 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ cluster: chr [1:6] "C1" "C2" "C3" "C3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ treatment          : chr "B"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">metadata</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sam"</span>, <span class="st">"trt"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="co">#The samples are nested in a condition.</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      trt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sam    A  B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A1  33  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A10 40  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A11 36  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A12 29  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A13 39  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A14 35  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A2  38  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A3  42  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A4  33  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A5  32  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A6  35  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A7  33  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A8  44  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   A9  44  0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B1   0 42</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B10  0 35</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B11  0 40</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B2   0 44</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B3   0 32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B4   0 47</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B5   0 42</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B6   0 50</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B7   0 40</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B8   0 63</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   B9   0 52</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Analyze differentially expressed (DE) genes specific to a cell-type using LMM.</span></span></span>
<span class="r-in"><span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="co">#expressions (log-transformed counts)</span></span></span>
<span class="r-in"><span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">libsize</span><span class="op">)</span> <span class="op">+</span> <span class="va">cls</span> <span class="op">+</span> <span class="va">cls</span><span class="op">:</span><span class="va">trt</span>, data <span class="op">=</span> <span class="va">dat</span><span class="op">$</span><span class="va">metadata</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">Z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">sam</span>, data <span class="op">=</span> <span class="va">dat</span><span class="op">$</span><span class="va">metadata</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Fitting LMM using cell-level data.</span></span></span>
<span class="r-in"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="lmmfit.html">lmmfit</a></span><span class="op">(</span><span class="va">Y</span>, <span class="va">X</span>, <span class="va">Z</span>, d <span class="op">=</span> <span class="va">d</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Fitting LMM using summary-level data.</span></span></span>
<span class="r-in"><span><span class="va">ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="sslmm.html">sslmm</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span>, <span class="va">Z</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fitss</span> <span class="op">&lt;-</span> <span class="fu"><a href="lmm.html">lmm</a></span><span class="op">(</span>summary.stats <span class="op">=</span> <span class="va">ss</span>, d <span class="op">=</span> <span class="va">d</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">fitss</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Hypothesis tests</span></span></span>
<span class="r-in"><span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="lmmtest.html">lmmtest</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       log(libsize)_t   clsC1_t   clsC2_t   clsC3_t   clsC4_t clsC1:trtB_t</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene1       3.556438 -2.415894 -2.595711 -2.557626 -2.446634    0.8697778</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene2       6.135313 -4.869716 -4.991892 -5.024973 -5.006548    0.3659413</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene3       3.609899 -2.312274 -2.246792 -2.148500 -2.220267    0.1515357</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene4       4.123944 -2.670994 -2.510841 -2.585615 -2.698308    0.9707563</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene5       2.818000 -1.322021 -1.238786 -1.253600 -1.205876   -0.9577914</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene6       3.592507 -2.552389 -2.766682 -2.699148 -2.635315    0.5414995</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       clsC2:trtB_t clsC3:trtB_t clsC4:trtB_t log(libsize)_p      clsC1_p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene1    2.0700456    1.5066385   -0.3263183   3.936946e-04 1.587661e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene2    1.0976568    1.5001771    1.6810047   1.226867e-09 1.300111e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene3   -0.1112106   -0.3659261   -0.4559326   3.216502e-04 2.096680e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene4    0.7423556    0.8883383    0.6462646   4.036515e-05 7.686618e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene5   -0.5997369   -1.0410003   -1.7515738   4.928509e-03 1.864664e-01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene6    1.5196520    0.3102291    0.6687761   3.436079e-04 1.084772e-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            clsC2_p      clsC3_p      clsC4_p clsC1:trtB_p clsC2:trtB_p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene1 9.579168e-03 1.068679e-02 1.459256e-02    0.3846325   0.03870667</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene2 7.060831e-07 5.971329e-07 6.556356e-07    0.7144869   0.27262100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene3 2.487275e-02 3.191586e-02 2.662620e-02    0.8795840   0.91147190</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene4 1.220264e-02 9.862323e-03 7.087769e-03    0.3319065   0.45804781</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene5 2.157181e-01 2.102832e-01 2.281532e-01    0.3384015   0.54881868</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene6 5.768368e-03 7.070027e-03 8.537187e-03    0.5882851   0.12891743</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       clsC3:trtB_p clsC4:trtB_p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene1    0.1322220   0.74425245</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene2    0.1338870   0.09307711</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene3    0.7144983   0.64853836</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene4    0.3745743   0.51825766</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene5    0.2981293   0.08015644</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene6    0.7564520   0.50379408</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#The DE genes specific to a cell-type.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">test</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">":"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        clsC1:trtB_t clsC2:trtB_t clsC3:trtB_t clsC4:trtB_t clsC1:trtB_p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene45    2.1720213   0.55889245   0.08637871   -0.9345875   0.03009058</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene46    0.8149981   2.28175377  -0.73193986    0.8343811   0.41526921</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene47    0.5897109   1.41986369   5.97111283    0.8070739   0.55551888</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene48    0.2007912  -0.06210722  -4.68109321   -2.2274365   0.84090305</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene49   -0.2895705   0.03168994  -3.22145147    0.9415348   0.77220544</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene50    1.8874002   1.58847538   6.53482944    1.2079394   0.05939855</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        clsC2:trtB_p clsC3:trtB_p clsC4:trtB_p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene45   0.57636139 9.311828e-01   0.35022860</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene46   0.02271521 4.643783e-01   0.40426735</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene47   0.15596186 3.279879e-09   0.41981754</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene48   0.95048997 3.250047e-06   0.02614204</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene49   0.97472569 1.316972e-03   0.34666032</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gene50   0.11249785 1.017003e-10   0.22735878</span>
<span class="r-in"><span><span class="co">#The true DE genes</span></span></span>
<span class="r-in"><span><span class="va">dat</span><span class="op">$</span><span class="va">DEgenes</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      gene       beta cluster</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 45 Gene45  0.5496463      C1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 46 Gene46  0.3652007      C2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 47 Gene47  0.8306937      C3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 48 Gene48 -0.9476624      C3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 49 Gene49 -0.8751216      C3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 50 Gene50  0.9050662      C3</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Changjiang Xu, Gary Bader.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

